File: 프로그래머스/2/12978. 배달/배달.kt
class Solution {
    lateinit var dist: Array<IntArray>
    
    fun solution(N: Int, road: Array<IntArray>, k: Int): Int {
        dist = Array(N + 1) { IntArray(N + 1) { 1_000_000_000 } }
        
        // 도로 시간 설정
        road.forEach {
            val (u, v, w) = it
            // 길이 여러 개면 더 짧은 길 선택
            dist[u][v] = minOf(dist[u][v], w)
            dist[v][u] = minOf(dist[v][u], w)
        }
        
        // 자기 자신은 0으로 초기화
        for (i in 1..N) {
            dist[i][i] = 0
        }
        
        // 플로이드 워셜
        for (m in 1..N) {
            for (i in 1..N) {
                for (j in 1..N) {
                    // 거쳐가는 게 바로 가는 것보다 짧은 경우 갱신
                    if (dist[i][m] + dist[m][j] < dist[i][j]) {
                        dist[i][j] = dist[i][m] + dist[m][j]
                    }
                }
            }
        }
        
        var answer = 0
        for (i in 1..N) {
            if (dist[1][i] <= k) {
                answer++
            }
        }
        return answer
    }
}

File: 프로그래머스/2/12978. 배달/배달.cpp
#include <vector>
#include <queue>

using namespace std;

int dist[51];
vector<pair<int, int>> graph[51];

void dijkstra() {
    // 최소힙 우선순위큐, pair<거리, 다음 노드 번호>
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    
    dist[1] = 0;
    pq.push({0, 1});
    
    while (!pq.empty()) {
        auto [curDist, curNode] = pq.top();
        pq.pop();
        
        if (curDist > dist[curNode]) continue;
        
        for (auto next : graph[curNode]) {
            int nextNode = next.second;
            int nextDist = next.first;
            
            if (dist[nextNode] > curDist + nextDist) {
                dist[nextNode] = curDist + nextDist;
                pq.push({dist[nextNode], nextNode});
            }
        }
    }
}

int solution(int N, vector<vector<int>> road, int K) {
    // input
    for (auto r : road) {
        int u = r[0], v = r[1], w = r[2];
        graph[u].push_back({w, v});
        graph[v].push_back({w, u});
    }
    
    // init dist
    for (int i = 1; i <= N; i++) dist[i] = 1e9;
    
    dijkstra();
    
    // count
    int answer = 0;
    for (int i = 1; i <= N; i++) {
        if (dist[i] <= K) answer++;
    }

    return answer;
}
